### ISSUE TITLE
CVEs alerts inventory for Vulnerability Detector - Exploratory testing

### ISSUE THREAD

**Issue by DProvinciani** [wazuh/wazuh#7928]
### Description

As part of the epic https://github.com/wazuh/wazuh/issues/7749, we are implementing a mechanism to have an inventory of vulnerabilities that affect the endpoints where the Wazuh agents are installed. Having said that, exploratory testing should be done over the full implementation of the epic.

This issue will be closed with a report of each of the participants of the exploratory testing, giving details of the findings as well as the fixes implemented if necessary.

**Comment by DProvinciani** [comment]
## Testing over Windows platform

**Test steps:**

1. Register the agent to the manager.
2. Check the `vuln_cves` table of the agent's database. (**EXPECTED**: `0` vulnerabilities)
3. Check the `alerts.json` file. (**EXPECTED**: `0` alerts)
4. Wait for the first Vulnerability Detector scan.
5. Check the `vuln_cves` table of the agent's database. (**EXPECTED**: `X` vulnerabilities)
6. Check the `alerts.json` file. (**EXPECTED**: `0` alerts)
7. Install a vulnerable package (`Wireshark 2.4.5 64-bit` as an example from the documentation)
8. Wait for the second Vulnerability Detector scan.
9. Check the `vuln_cves` table of the agent's database. (**EXPECTED**: `Y` vulnerabilities)
10. Check the `alerts.json` file. (**EXPECTED**: `Z` alerts where `Z=Y-X`)
11. Uninstall the vulnerable package.
12. Wait for the third Vulnerability Detector scan.
13. Check the `vuln_cves` table of the agent's database. (**EXPECTED**: `X` vulnerabilities)
14. Check the `alerts.json` file. (**EXPECTED**: `2*Z` alerts where `Z=Y-X`. This is because of the CVE removal alerts).

### Testing results

|System|Platform|Result|
|---|---|---|
|Windows 7|x64|**PASS**|
|Windows 8|x64|**PASS**|
|Windows 8.1|x64|**PASS**|
|Windows 10 RS1|x64|**PASS**|

### Testing example (Windows 7 X64)

Before the first scan

```
[root@localhost ~]# python3 wdb-query.py 001 "SELECT COUNT(*) FROM vuln_cves"
[
    {
        "COUNT(*)": 0
    }
]
[root@localhost ~]# grep vulnerability-detector /var/ossec/logs/alerts/alerts.json | wc -l
0
```

After the first scan (`BASELINE`)

```
[root@localhost ~]# python3 wdb-query.py 001 "SELECT COUNT(*) FROM vuln_cves"
[
    {
        "COUNT(*)": 1162
    }
]
[root@localhost ~]# grep vulnerability-detector /var/ossec/logs/alerts/alerts.json | wc -l
0
```

After installing `Wireshark 2.4.5 64-bit`

```
[root@localhost ~]# python3 wdb-query.py 001 "SELECT COUNT(*) FROM vuln_cves"
[
    {
        "COUNT(*)": 1224
    }
]
[root@localhost ~]# grep vulnerability-detector /var/ossec/logs/alerts/alerts.json | wc -l
62
```

After uninstalling `Wireshark 2.4.5 64-bit`

```
[root@localhost ~]# python3 wdb-query.py 001 "SELECT COUNT(*) FROM vuln_cves"
[
    {
        "COUNT(*)": 1162
    }
]
[root@localhost ~]# grep vulnerability-detector /var/ossec/logs/alerts/alerts.json | wc -l
124
```

**Comment by pereyra-m** [comment]
### Tests

Some manual tests were performed in a this environment:
- Manager CentOS 7. Wazuh V4.2 and V4.3. The manager was scanned also as an agent.
- Agents:
    - Ubuntu 20.10 Wazuh V4.1.2
    - Windows 2008 Wazuh V4.1.2

These were the conclusions from the tests:
- The alerts generated with the V4.2 and V4.3 were the same: the analysis engines works the same way
- The reference from `sys_osinfo/sys_programs` tables was correctly propagated  to the `vuln_cves` table
- The LAST_PARTIAL_SCAN and LAST_FULL_SCAN timestamps are updated correctly
- The BASELINE scan doesn't produce alerts
- The new alert for removed vulnerabilities works as expected
- The PARTIAL scan also removes the OBSOLETE vulnerabilities
- A hotfix insertion/removal triggers an OS scan
- The FULL scan is triggered only after a `min_full_scan_interval` and a feed update
- After an upgrade, the removed configuration tag is alerted as expected
- After a source upgrade, the vulnerabilities table is cleared but the old `cve.db` DB isn't removed. This upgrade mechanism issue isn't related to this epic and was reported in #8214 
- If a vulnerable package is removed during a scan, it isn't inserted at all. This behavior could be different and will be discussed with the team

**Comment by palaciosjeremias** [comment]
**Environment:**
- Manager 4.2 as the baseline
- Manager _dev-cves-alerts-inventory_ as target
- Agent Ubuntu 20.04 
- Agent Windows 10 1909

**Tests:**
Triggers:
- Connect an agent for the first time and enable vuln_detector. Expected result: A baseline scan triggers.
- Wait **interval**. Expected result: A partial scan triggers.
- Wait **min_full_scan_interval** without a feed update. Expected result: A partial scan triggers
- Update the feed without waiting enough for **min_full_scan_interval**.  Expected result: A partial scan triggers
- Wait **min_full_scan_interval** and update the OS feed.  Expected result: A full scan triggers.
- Wait **min_full_scan_interval** and update the NVD feed.  Expected result: A fullscan triggers.
|Agent||Result|
|---|---|
|Windows 10 1909|**PASS**|
|Ubuntu 20.04|**PASS**|

Vulnerabilities on vuln_cve:
- Run a baseline scan. Expected result: Same vulnerabilities detected as the baseline (4.2).
- Remove a Vulnerable package (Wireshark 3.0.6) and wait for a partial scan. Expected result: The respective entries in vuln_cve are eliminated.
- Add a vulnerable package and wait for a partial scan. Expected result: The respective entries in vuln_cve are added.
- Remove the vulnerability from vuln_cve and run a full scan. Expected result All the initial entries are regenerated.
|Agent||Result|
|---|---|
|Windows 10 1909|**PASS**|

Alerts:
- Run a baseline scan. Expected result: No alerts are generated.
- Remove a vulnerable package and wait for a partial or full scan. Expected result: A vulnerability removal alert is generated.
- Add a vulnerable package and wait for a partial or full scan. Expected result: A new vulnerability alert is generated.
- Wait for a partial scan without any package modification. Expected result: No alerts are generated.
|Windows 10 1909|**PASS**|
|Ubuntu 20.04|**PASS**|

**Comment by DProvinciani** [comment]
Having the results of the exploratory documented. We close the issue.

**Comment by DProvinciani** [comment]
Considering the multiples rebase that the branch of this epic received because of the delay for the merge to master, we reopen this issue in order to run again an exploratory testing over the whole epic implementation.

**Comment by MiguelazoDS** [comment]
## Testing on Centos8 manager, Ubuntu bionic and Windows 2019 agents.
----

### Environment:
- Centos8 manager Wazuh 4.3 [branch](https://github.com/wazuh/wazuh/tree/dev-8529-vul-det-atomic-query)
- Ubuntu Bionic Wazuh 4.1.5 (package installation)
- Windows 2019 server Wazuh 4.1.5 (package installation)

### Tests Centos8 manager without agents:
----
- After enabling the vulnerability detector the `Baseline scan` is launched. No alerts generated.
![01- baseline manager](https://user-images.githubusercontent.com/13010397/119001887-c3edb080-b962-11eb-888c-4b6cc264fa86.png)
- `Partial scans` run every minute (<interval>1m</interval>). On this example at **19:23:55** and again at **19:24:55**.
![02- partial scan manager after baseline (1 min interval)](https://user-images.githubusercontent.com/13010397/119002374-3199dc80-b963-11eb-8e98-f58b1f3ba6de.png)
- Force `full scan` deleting `cve.db` to get new feed information as well as change the feed update and min_full_scan_interval settings to 1m.
![04- full scan ](https://user-images.githubusercontent.com/13010397/119002756-863d5780-b963-11eb-8a4c-11ac533b2ea6.png)
- After the `full scan` only `partial scans` are executed.
![05- only partial scan after the full scan](https://user-images.githubusercontent.com/13010397/119002970-b8e75000-b963-11eb-85ed-e22cd6eff9a6.png)
### Tests adding Ubuntu bionic agent:
----
- `Baseline scan` runs after the agent enrollment occurs. No alerts generated.
  <img src="https://user-images.githubusercontent.com/13010397/119003214-ee8c3900-b963-11eb-93ba-c79d3b5a51c5.png" alt="drawing" width="400"/>
![07- baseline starts for ubuntu agent](https://user-images.githubusercontent.com/13010397/119003226-f055fc80-b963-11eb-9f63-8daf3dd877d1.png)
- Operating system update results in vulnerabilities fixed. Removal alerts are triggered.
```json
{
  "timestamp": "2021-05-19T20:32:35.953-0300",
  "rule": {
    "level": 3,
    "description": "The CVE-2020-27749 that affected grub2-common was eliminated due to a package removal/update or a system upgrade",
    "id": "23502",
    "firedtimes": 29,
    "mail": false,
    "groups": [
      "vulnerability-detector"
    ],
    "gdpr": [
      "IV_35.7.d"
    ],
    "pci_dss": [
      "11.2.1",
      "11.2.3"
    ],
    "tsc": [
      "CC7.1",
      "CC7.2"
    ]
  },
  "agent": {
    "id": "001",
    "name": "ubuntu18",
    "ip": "10.0.2.15"
  },
  "manager": {
    "name": "centos8"
  },
  "id": "1621467155.742606",
  "decoder": {
    "name": "json"
  },
  "data": {
    "vulnerability": {
      "package": {
        "name": "grub2-common",
        "version": "2.02-2ubuntu8.21",
        "architecture": "amd64"
      },
      "cve": "CVE-2020-27749",
      "status": "REMOVED",
      "reference": "759c21ce434cd2960cf332bc14556cdaa04d10d1",
      "type": "PACKAGE",
      "title": "CVE-2020-27749 affecting grub2-common was eliminated"
    }
  },
  "location": "vulnerability-detector"
}
```
- Vulnerabilities number increased in Ubuntu agent after installing qemu package.
  <img src="https://user-images.githubusercontent.com/13010397/119004946-61e27a80-b965-11eb-96c2-11fd6f8c3569.png" alt="drawing" width="450"/>
  <img src="https://user-images.githubusercontent.com/13010397/119005744-1c727d00-b966-11eb-99a0-2fda1ee8ffa0.png" alt="drawing" width="550"/>
  <img src="https://user-images.githubusercontent.com/13010397/119005665-04026280-b966-11eb-8371-5c681823edfa.png" alt="drawing" width="650"/>
- One of the new vulnerabilities added due to Qemu package installation.
 ```json
  {
  "timestamp": "2021-05-20T11:19:08.251-0300",
  "rule": {
    "level": 5,
    "description": "CVE-2021-20196 affects libxen-4.9",
    "id": "23503",
    "firedtimes": 201,
    "mail": false,
    "groups": [
      "vulnerability-detector"
    ],
    "gdpr": [
      "IV_35.7.d"
    ],
    "pci_dss": [
      "11.2.1",
      "11.2.3"
    ],
    "tsc": [
      "CC7.1",
      "CC7.2"
    ]
  },
  "agent": {
    "id": "001",
    "name": "ubuntu18",
    "ip": "10.0.2.15"
  },
  "manager": {
    "name": "centos8"
  },
  "id": "1621520348.3649209",
  "decoder": {
    "name": "json"
  },
  "data": {
    "vulnerability": {
      "package": {
        "name": "libxen-4.9",
        "source": "xen",
        "version": "4.9.2-0ubuntu1",
        "architecture": "amd64",
        "condition": "Package unfixed"
      },
      "cve": "CVE-2021-20196",
      "title": "CVE-2021-20196 affects libxen-4.9",
      "rationale": "block: fdc: null pointer dereference may lead to guest crash",
      "severity": "Low",
      "published": "2021-01-25T00:00:00Z",
      "bugzilla_references": [
        "https://bugs.launchpad.net/qemu/+bug/1912780",
        "https://bugzilla.redhat.com/show_bug.cgi?id=1919210"
      ],
      "references": [
        "http://people.canonical.com/~ubuntu-security/cve/2021/CVE-2021-20196.html",
        "https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20196",
        "https://lists.nongnu.org/archive/html/qemu-devel/2021-01/msg05986.html",
        "https://www.openwall.com/lists/oss-security/2021/01/28/1"
      ]
    }
  },
  "location": "vulnerability-detector"
}
```
- Removed Qemu package results in Removal alerts.
```json
{
  "timestamp": "2021-05-20T11:04:01.773-0300",
  "rule": {
    "level": 3,
    "description": "The CVE-2021-3544 that affected qemu-user was eliminated due to a package removal/update or a system upgrade",
    "id": "23502",
    "firedtimes": 314,
    "mail": false,
    "groups": [
      "vulnerability-detector"
    ],
    "gdpr": [
      "IV_35.7.d"
    ],
    "pci_dss": [
      "11.2.1",
      "11.2.3"
    ],
    "tsc": [
      "CC7.1",
      "CC7.2"
    ]
  },
  "agent": {
    "id": "001",
    "name": "ubuntu18",
    "ip": "10.0.2.15"
  },
  "manager": {
    "name": "centos8"
  },
  "id": "1621519441.2499656",
  "decoder": {
    "name": "json"
  },
  "data": {
    "vulnerability": {
      "package": {
        "name": "qemu-user",
        "version": "1:2.11+dfsg-1ubuntu7.36",
        "architecture": "amd64"
      },
      "cve": "CVE-2021-3544",
      "status": "REMOVED",
      "reference": "022090ef8e3563655179d619c5dabd30dabe39fc",
      "type": "PACKAGE",
      "title": "CVE-2021-3544 affecting qemu-user was eliminated"
    }
  },
  "location": "vulnerability-detector"
}
```
### Tests adding Windows 2019 Server agent:
----
- Status of agents table in Global database.
![16- agents table in global db](https://user-images.githubusercontent.com/13010397/119006898-26e14680-b967-11eb-89e6-527b8d2bcc07.png)
- `Baseline scan` runs after the agent enrollment occurs. No alerts generated
![17- baseline for windows](https://user-images.githubusercontent.com/13010397/119006821-14ffa380-b967-11eb-9b69-4d371352e20c.png)
- Hotfixes listed in `sys_hotfixes` table.
  <img src="https://user-images.githubusercontent.com/13010397/119007189-6445d400-b967-11eb-83f6-a43909881a44.png" alt="drawing" width="500"/>
- New hotfixes after update.
  <img src="https://user-images.githubusercontent.com/13010397/119007222-6a3bb500-b967-11eb-8f18-5e3afbd1270d.png" alt="drawing" width="500"/>
- Removal alerts due to new security updates installation.
```json
{
  "timestamp": "2021-05-20T10:06:25.579-0300",
  "rule": {
    "level": 3,
    "description": "The CVE-2021-31194 that affected Windows Server 2019 was eliminated due to a package removal/update or a system upgrade",
    "id": "23502",
    "firedtimes": 1,
    "mail": false,
    "groups": [
      "vulnerability-detector"
    ],
    "gdpr": [
      "IV_35.7.d"
    ],
    "pci_dss": [
      "11.2.1",
      "11.2.3"
    ],
    "tsc": [
      "CC7.1",
      "CC7.2"
    ]
  },
  "agent": {
    "id": "002",
    "name": "WIN-MCUEPDU1M6L",
    "ip": "10.0.2.15"
  },
  "manager": {
    "name": "centos8"
  },
  "id": "1621515985.652678",
  "decoder": {
    "name": "json"
  },
  "data": {
    "vulnerability": {
      "package": {
        "name": "Windows Server 2019"
      },
      "cve": "CVE-2021-31194",
      "status": "REMOVED",
      "reference": "3b7b264c19ea59e5ef86dae0b1463f5c42ff7d91",
      "type": "OS",
      "title": "CVE-2021-31194 affecting Windows Server 2019 was eliminated"
    }
  },
  "location": "vulnerability-detector"
}
```
- Uninstalling security update `KB5003171`
  <img src="https://user-images.githubusercontent.com/13010397/119007507-aa029c80-b967-11eb-9f6c-566244114606.png" alt="drawing" width="500"/>
- Vulnerability alerts after KB deletion. `"condition": "KB5003171 patch is not installed"`
```json
{
  "timestamp": "2021-05-20T10:54:29.612-0300",
  "rule": {
    "level": 10,
    "description": "CVE-2021-31194 affects Windows Server 2019",
    "id": "23505",
    "firedtimes": 2,
    "mail": false,
    "groups": [
      "vulnerability-detector"
    ],
    "gdpr": [
      "IV_35.7.d"
    ],
    "pci_dss": [
      "11.2.1",
      "11.2.3"
    ],
    "tsc": [
      "CC7.1",
      "CC7.2"
    ]
  },
  "agent": {
    "id": "002",
    "name": "WIN-MCUEPDU1M6L",
    "ip": "10.0.2.15"
  },
  "manager": {
    "name": "centos8"
  },
  "id": "1621518869.1095511",
  "decoder": {
    "name": "json"
  },
  "data": {
    "vulnerability": {
      "package": {
        "name": "Windows Server 2019",
        "generated_cpe": "o:microsoft:windows_server_2019:1809:::::::",
        "condition": "KB5003171 patch is not installed"
      },
      "cvss": {
        "cvss2": {
          "vector": {
            "attack_vector": "network",
            "access_complexity": "low",
            "authentication": "single",
            "confidentiality_impact": "partial",
            "integrity_impact": "partial",
            "availability": "partial"
          },
          "base_score": "6.500000"
        },
        "cvss3": {
          "vector": {
            "attack_vector": "network",
            "access_complexity": "low",
            "privileges_required": "low",
            "user_interaction": "none",
            "scope": "unchanged",
            "confidentiality_impact": "high",
            "integrity_impact": "high",
            "availability": "high"
          },
          "base_score": "8.800000"
        }
      },
      "cve": "CVE-2021-31194",
      "title": "CVE-2021-31194 affects Windows Server 2019",
      "rationale": "OLE Automation Remote Code Execution Vulnerability",
      "severity": "High",
      "published": "2021-05-11",
      "updated": "2021-05-17",
      "cwe_reference": "NVD-CWE-noinfo",
      "references": [
        "https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2021-31194",
        "https://nvd.nist.gov/vuln/detail/CVE-2021-31194"
      ],
      "assigner": "secure@microsoft.com",
      "cve_version": "4.0"
    }
  },
  "location": "vulnerability-detector"
}
```

**Comment by pereyra-m** [comment]
### Exploratory tests

The manual tests were performed with the following environment:
- Manager Ubuntu 20.04 **v4.1.5** -> **v4.3.0**
- Agent Windows Server 2008  **v4.1.5** 
- Agent Arch Linux (Manjaro) **v4.1.5** -> **v4.2.0**

The vulnerabilities are reported as expected. There were only two situations found related to the Arch Linux agent:
- The OS name is not correctly extracted by the **v4.1.5** agent. This is fixed in #8703 
- The agent's packages can't be inserted if the manager has been updated because the **CHECK** constraint of the `sys_programs` table doesn't recognize the **pacman** string. This doesn't happen in a clean installation.

**Comment by DProvinciani** [comment]
Considering the validations that we made, we close the issue.
